File: backend/school_service/api/routes/streams.py
"""API routes for Stream management."""

from fastapi import APIRouter, HTTPException, status, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional

from school_service.core.database import get_db
from school_service.services.stream_service import StreamService
from school_service.services.class_service import ClassService
from shared.schemas.stream_schema import (
    StreamCreate,
    StreamUpdate,
    StreamResponse,
)
from shared.schemas.user import UserResponse
from school_service.api.routes.auth import get_current_user

router = APIRouter(prefix="/api/v1/streams", tags=["streams"])


@router.post(
    "",
    response_model=StreamResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new stream",
    description="""
    Create a new stream within a class.
    
    The stream name must be unique within the class.
    The class must belong to the authenticated user's school.
    """,
    responses={
        201: {
            "description": "Stream created successfully",
            "model": StreamResponse,
        },
        404: {
            "description": "Class not found",
        },
        409: {
            "description": "Stream name already exists",
            "content": {
                "application/json": {
                    "example": {
                        "detail": "Stream name 'A' already exists for this class"
                    }
                }
            },
        },
        422: {
            "description": "Validation error",
        },
    },
)
async def create_stream(
    stream_data: StreamCreate,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Create a new stream within a class.
    
    - **Authentication**: Required (JWT token)
    - **Authorization**: Class must belong to user's school
    """
    stream_service = StreamService(db)
    class_service = ClassService(db)
    
    # Verify class exists and belongs to user's school
    class_obj = await class_service.get_class_by_id(
        class_id=stream_data.class_id,
        school_id=current_user.school_id,
    )
    
    if not class_obj:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Class not found",
        )
    
    try:
        stream = await stream_service.create_stream(stream_data)
        return StreamResponse.model_validate(stream)
    except ValueError as e:
        error_msg = str(e)
        if "name" in error_msg.lower() and ("unique" in error_msg.lower() or "already exists" in error_msg.lower()):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=error_msg,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_msg,
        )


@router.get(
    "",
    response_model=List[StreamResponse],
    summary="List streams",
    description="""
    List streams, optionally filtered by class.
    
    Only returns streams from classes in the authenticated user's school.
    """,
    responses={
        200: {
            "description": "List of streams retrieved successfully",
            "model": List[StreamResponse],
        },
        404: {
            "description": "Class not found (if class_id provided)",
        },
    },
)
async def list_streams(
    class_id: Optional[int] = Query(None, description="Filter by class ID"),
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    List streams, optionally filtered by class.
    
    - **Authentication**: Required (JWT token)
    - **Authorization**: Only returns streams from user's school classes
    """
    stream_service = StreamService(db)
    class_service = ClassService(db)
    
    # If class_id provided, verify it belongs to user's school
    if class_id:
        class_obj = await class_service.get_class_by_id(
            class_id=class_id,
            school_id=current_user.school_id,
        )
        if not class_obj:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Class not found",
            )
    
    streams = await stream_service.list_streams(
        school_id=current_user.school_id,
        class_id=class_id,
    )
    return [StreamResponse.model_validate(s) for s in streams]


@router.get(
    "/{stream_id}",
    response_model=StreamResponse,
    summary="Get stream by ID",
    description="""
    Get a specific stream by ID.
    
    The stream must belong to a class in the authenticated user's school.
    """,
    responses={
        200: {
            "description": "Stream retrieved successfully",
            "model": StreamResponse,
        },
        404: {
            "description": "Stream not found",
            "content": {
                "application/json": {
                    "example": {
                        "detail": "Stream not found"
                    }
                }
            },
        },
    },
)
async def get_stream(
    stream_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Get a stream by ID.
    
    - **Authentication**: Required (JWT token)
    - **Authorization**: Stream must belong to user's school (via class)
    """
    stream_service = StreamService(db)
    
    stream = await stream_service.get_stream_by_id(
        stream_id=stream_id,
        school_id=current_user.school_id,
    )
    
    if not stream:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stream not found",
        )
    
    return StreamResponse.model_validate(stream)


@router.put(
    "/{stream_id}",
    response_model=StreamResponse,
    summary="Update stream",
    description="""
    Update stream information.
    
    Class ID cannot be changed.
    """,
    responses={
        200: {
            "description": "Stream updated successfully",
            "model": StreamResponse,
        },
        404: {
            "description": "Stream not found",
            "content": {
                "application/json": {
                    "example": {
                        "detail": "Stream not found"
                    }
                }
            },
        },
        409: {
            "description": "Stream name already exists",
        },
    },
)
async def update_stream(
    stream_id: int,
    stream_data: StreamUpdate,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Update a stream.
    
    - **Authentication**: Required (JWT token)
    - **Authorization**: Stream must belong to user's school (via class)
    - **Immutable Fields**: class_id
    """
    stream_service = StreamService(db)
    
    # Verify stream exists and belongs to user's school
    existing_stream = await stream_service.get_stream_by_id(
        stream_id=stream_id,
        school_id=current_user.school_id,
    )
    
    if not existing_stream:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stream not found",
        )
    
    try:
        updated_stream = await stream_service.update_stream(
            stream_id=stream_id,
            stream_data=stream_data,
            school_id=current_user.school_id,
        )
        
        if not updated_stream:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Stream not found",
            )
        
        return StreamResponse.model_validate(updated_stream)
    except ValueError as e:
        error_msg = str(e)
        if "name" in error_msg.lower() and ("unique" in error_msg.lower() or "already exists" in error_msg.lower()):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=error_msg,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_msg,
        )


@router.delete(
    "/{stream_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete stream (soft delete)",
    description="""
    Soft delete a stream (deactivate).
    
    The stream data is preserved but marked as deleted.
    This is a soft delete - the stream record remains in the database.
    """,
    responses={
        204: {
            "description": "Stream deleted successfully",
        },
        404: {
            "description": "Stream not found",
            "content": {
                "application/json": {
                    "example": {
                        "detail": "Stream not found"
                    }
                }
            },
        },
    },
)
async def delete_stream(
    stream_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Soft delete a stream.
    
    - **Authentication**: Required (JWT token)
    - **Authorization**: Stream must belong to user's school (via class)
    - **Operation**: Soft delete (sets is_deleted = true)
    """
    stream_service = StreamService(db)
    
    # Verify stream exists and belongs to user's school
    existing_stream = await stream_service.get_stream_by_id(
        stream_id=stream_id,
        school_id=current_user.school_id,
    )
    
    if not existing_stream:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stream not found",
        )
    
    deleted = await stream_service.delete_stream(
        stream_id=stream_id,
        school_id=current_user.school_id,
    )
    
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stream not found",
        )
    
    return None

File: frontend/lib/api/streams.ts
/**
 * API client functions for Stream management.
 * 
 * Base URL: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8001'
 * Endpoints:
 * - GET /api/v1/streams - List streams (optionally filtered by class)
 * - GET /api/v1/streams/{id} - Get stream by ID
 * - POST /api/v1/streams - Create a new stream
 * - PUT /api/v1/streams/{id} - Update stream
 * - DELETE /api/v1/streams/{id} - Soft delete stream
 */

import axios from 'axios';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8001';

/**
 * Stream response type from the API.
 */
export interface StreamResponse {
  id: number;
  class_id: number;
  name: string;
  description: string | null;
  is_deleted: boolean;
  created_at: string; // ISO datetime
  updated_at: string | null;
}

/**
 * Stream creation data.
 */
export interface StreamCreateData {
  class_id: number;
  name: string;
  description?: string | null;
}

/**
 * Stream update data.
 */
export interface StreamUpdateData {
  name?: string;
  description?: string | null;
}

/**
 * API error response structure.
 */
export interface ApiError {
  detail: string | Array<{ loc: string[]; msg: string; type: string }>;
}

/**
 * Custom error class for Stream API errors.
 */
export class StreamApiError extends Error {
  constructor(
    message: string,
    public statusCode: number,
    public fieldErrors?: Record<string, string>
  ) {
    super(message);
    this.name = 'StreamApiError';
  }
}

/**
 * List streams, optionally filtered by class.
 * 
 * @param token - JWT authentication token
 * @param classId - Optional class ID to filter by
 * @returns Promise resolving to list of streams
 * @throws StreamApiError if request fails
 */
export async function listStreams(
  token: string,
  classId?: number | null
): Promise<StreamResponse[]> {
  try {
    const queryParams = new URLSearchParams();
    if (classId) {
      queryParams.append('class_id', classId.toString());
    }

    const url = classId
      ? `${API_BASE_URL}/api/v1/streams?${queryParams.toString()}`
      : `${API_BASE_URL}/api/v1/streams`;

    const response = await axios.get<StreamResponse[]>(
      url,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        validateStatus: (status) => status < 500,
      }
    );

    if (response.status >= 400) {
      const errorData = response.data as unknown as ApiError | undefined;
      const statusCode = response.status;

      if (statusCode === 401) {
        throw new StreamApiError(
          'Authentication required. Please log in and try again.',
          statusCode
        );
      }

      if (statusCode === 404) {
        throw new StreamApiError('Class not found', statusCode);
      }

      const message = typeof errorData?.detail === 'string'
        ? errorData.detail
        : `Failed to load streams (${statusCode})`;
      throw new StreamApiError(message, statusCode);
    }

    return response.data;
  } catch (error) {
    if (error instanceof StreamApiError) {
      throw error;
    }
    if (axios.isAxiosError(error)) {
      throw new StreamApiError(
        error.message || 'Failed to load streams',
        error.response?.status || 500
      );
    }
    throw new StreamApiError('An unexpected error occurred', 500);
  }
}

/**
 * Get a stream by ID.
 * 
 * @param token - JWT authentication token
 * @param streamId - Stream ID
 * @returns Promise resolving to stream data
 * @throws StreamApiError if request fails
 */
export async function getStream(
  token: string,
  streamId: number
): Promise<StreamResponse> {
  try {
    const response = await axios.get<StreamResponse>(
      `${API_BASE_URL}/api/v1/streams/${streamId}`,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        validateStatus: (status) => status < 500,
      }
    );

    if (response.status >= 400) {
      const errorData = response.data as unknown as ApiError | undefined;
      const statusCode = response.status;

      if (statusCode === 401) {
        throw new StreamApiError(
          'Authentication required. Please log in and try again.',
          statusCode
        );
      }

      if (statusCode === 404) {
        throw new StreamApiError('Stream not found', statusCode);
      }

      const message = typeof errorData?.detail === 'string'
        ? errorData.detail
        : `Failed to load stream (${statusCode})`;
      throw new StreamApiError(message, statusCode);
    }

    return response.data;
  } catch (error) {
    if (error instanceof StreamApiError) {
      throw error;
    }
    if (axios.isAxiosError(error)) {
      throw new StreamApiError(
        error.message || 'Failed to load stream',
        error.response?.status || 500
      );
    }
    throw new StreamApiError('An unexpected error occurred', 500);
  }
}

/**
 * Create a new stream.
 * 
 * @param token - JWT authentication token
 * @param data - Stream creation data
 * @returns Promise resolving to created stream
 * @throws StreamApiError if request fails
 */
export async function createStream(
  token: string,
  data: StreamCreateData
): Promise<StreamResponse> {
  try {
    const response = await axios.post<StreamResponse>(
      `${API_BASE_URL}/api/v1/streams`,
      data,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        validateStatus: (status) => status < 500,
      }
    );

    if (response.status >= 400) {
      const errorData = response.data as unknown as ApiError | undefined;
      const statusCode = response.status;

      if (statusCode === 401) {
        throw new StreamApiError(
          'Authentication required. Please log in and try again.',
          statusCode
        );
      }

      if (statusCode === 404) {
        throw new StreamApiError('Class not found', statusCode);
      }

      if (statusCode === 409) {
        const message = typeof errorData?.detail === 'string'
          ? errorData.detail
          : 'Stream name already exists for this class';
        throw new StreamApiError(message, statusCode);
      }

      // Handle validation errors
      if (statusCode === 422 && Array.isArray(errorData?.detail)) {
        const fieldErrors: Record<string, string> = {};
        errorData.detail.forEach((err) => {
          if (err.loc && err.loc.length > 0) {
            const field = err.loc[err.loc.length - 1];
            fieldErrors[field] = err.msg;
          }
        });
        throw new StreamApiError(
          'Validation failed. Please check your input.',
          statusCode,
          fieldErrors
        );
      }

      const message = typeof errorData?.detail === 'string'
        ? errorData.detail
        : `Failed to create stream (${statusCode})`;
      throw new StreamApiError(message, statusCode);
    }

    return response.data;
  } catch (error) {
    if (error instanceof StreamApiError) {
      throw error;
    }
    if (axios.isAxiosError(error)) {
      throw new StreamApiError(
        error.message || 'Failed to create stream',
        error.response?.status || 500
      );
    }
    throw new StreamApiError('An unexpected error occurred', 500);
  }
}

/**
 * Update a stream.
 * 
 * @param token - JWT authentication token
 * @param streamId - Stream ID
 * @param data - Stream update data
 * @returns Promise resolving to updated stream
 * @throws StreamApiError if request fails
 */
export async function updateStream(
  token: string,
  streamId: number,
  data: StreamUpdateData
): Promise<StreamResponse> {
  try {
    const response = await axios.put<StreamResponse>(
      `${API_BASE_URL}/api/v1/streams/${streamId}`,
      data,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        validateStatus: (status) => status < 500,
      }
    );

    if (response.status >= 400) {
      const errorData = response.data as unknown as ApiError | undefined;
      const statusCode = response.status;

      if (statusCode === 401) {
        throw new StreamApiError(
          'Authentication required. Please log in and try again.',
          statusCode
        );
      }

      if (statusCode === 404) {
        throw new StreamApiError('Stream not found', statusCode);
      }

      if (statusCode === 409) {
        const message = typeof errorData?.detail === 'string'
          ? errorData.detail
          : 'Stream name already exists for this class';
        throw new StreamApiError(message, statusCode);
      }

      // Handle validation errors
      if (statusCode === 422 && Array.isArray(errorData?.detail)) {
        const fieldErrors: Record<string, string> = {};
        errorData.detail.forEach((err) => {
          if (err.loc && err.loc.length > 0) {
            const field = err.loc[err.loc.length - 1];
            fieldErrors[field] = err.msg;
          }
        });
        throw new StreamApiError(
          'Validation failed. Please check your input.',
          statusCode,
          fieldErrors
        );
      }

      const message = typeof errorData?.detail === 'string'
        ? errorData.detail
        : `Failed to update stream (${statusCode})`;
      throw new StreamApiError(message, statusCode);
    }

    return response.data;
  } catch (error) {
    if (error instanceof StreamApiError) {
      throw error;
    }
    if (axios.isAxiosError(error)) {
      throw new StreamApiError(
        error.message || 'Failed to update stream',
        error.response?.status || 500
      );
    }
    throw new StreamApiError('An unexpected error occurred', 500);
  }
}

/**
 * Delete a stream (soft delete).
 * 
 * @param token - JWT authentication token
 * @param streamId - Stream ID
 * @throws StreamApiError if request fails
 */
export async function deleteStream(
  token: string,
  streamId: number
): Promise<void> {
  try {
    const response = await axios.delete(
      `${API_BASE_URL}/api/v1/streams/${streamId}`,
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        validateStatus: (status) => status < 500,
      }
    );

    if (response.status >= 400) {
      const errorData = response.data as unknown as ApiError | undefined;
      const statusCode = response.status;

      if (statusCode === 401) {
        throw new StreamApiError(
          'Authentication required. Please log in and try again.',
          statusCode
        );
      }

      if (statusCode === 404) {
        throw new StreamApiError('Stream not found', statusCode);
      }

      const message = typeof errorData?.detail === 'string'
        ? errorData.detail
        : `Failed to delete stream (${statusCode})`;
      throw new StreamApiError(message, statusCode);
    }
  } catch (error) {
    if (error instanceof StreamApiError) {
      throw error;
    }
    if (axios.isAxiosError(error)) {
      throw new StreamApiError(
        error.message || 'Failed to delete stream',
        error.response?.status || 500
      );
    }
    throw new StreamApiError('An unexpected error occurred', 500);
  }
}

